t1535 - Adds functionality that allows some restrictions text and special instructions text to be defined in config against the rights category and then displayed accordingly in metadata panel

AndyKilmory opened the pull request:
## What does this change do?

**Summary**
This change allows some Restrictions text and Special Instructions text to be defined in the Application level config agianst specific Rights Categories. This information will then be displayed in the metadata and info panels when the image is selected (either as part of a set of images or individually). The information will displayed alongside any restrictions or special intructions that are added by the user and are specific to the image in question.

The configuration derived restrictions and special instructions will not be editable by the user in either the info panel or the metadata panel. If the rights category is updated then the any restrictions and special instructions derived from configuration will be displayed in place of those associated with the prior rights category.

Configuration derived restrictions and special instructions will also be displayed on the Upload page during the editing of the image metadata during the upload process. Thus the uploader will be able to see which restrictions and special instructions will be displayed with this image due to the selected rights category.

It also changes the behaviour of the Special Instructions group in the info panel and metadata editor - previously if an image didn't have any special instructions this section was not visible - this made it hard to add special instructions to an image that had not got any. This PR changes that behaviour so that the Special Instruction group is visible even if the image does not contain any value in that field. The behaviour now closely mirrors that of other metadata fields like Description wherer the User if told the value is unknown and given the option to edit if they have edit rights - see below...

![Screenshot 2024-07-01 at 10 36 33](https://github.com/guardian/grid/assets/128470622/3edcf324-fba7-499a-a25f-2a5cc816fac1)


**Defining Special Intructions and Restrictions in Config**
The rights category Special Instructions and Restrictions should be defined in the application configuration as json objects according to the patterns;

usageInstructions {
  {{rights category key}} = "{{rights derived special instructions}}"
}

usageRestrictions {
  {{rights category key}} = "{{rights derived restrictions}}"
}

for example;

usageInstructions {
    staff-photographer = "You must also make sure that any use meets BBC’s Editorial Guidelines."
    contract-photographer = "This image can only be used for non-commercial BBC promotional activity of the programme from which it came."
}

usageRestrictions {
    staff-photographer = "See the special instructions related to this image and consider if the use is editorially justified"
    contract-photographer = "See the special instructions and lease details related to this image."
    obituary = "Only to be used in conjunction with named person."
}  

If the _usageInstructions_ and _usageRestrictions_ objects are empty or absent from the configuration then the code will continue to operate as normal with no additional information displayed in conjunction with the images.

**Display of information in grid and image views**
In main image grid the additional information will appear as follows;

![Screenshot 2024-06-05 at 10 42 04](https://github.com/guardian/grid/assets/128470622/78cfc204-9259-4781-865f-5c92d7540dd1)

When an image is selected the information will appear as;

![Screenshot 2024-06-05 at 10 43 56](https://github.com/guardian/grid/assets/128470622/e1d0771d-28ff-450b-8c2b-b34860eafe3f)

Note: This image also shows how the user is able to add further special instructions that are image specific in addition to those derived from the rights category.

**Display on Uploads Page**
On the uploads page the config declared rights restrictions and special instructions will appear as;

![Screenshot 2024-06-05 at 10 49 23](https://github.com/guardian/grid/assets/128470622/ce2941ce-2254-474e-aa0b-256eced0a4fc)

This eaxmple shows both restrictions and special instructions drawn from config. Note the special instructions element will only be updated and displayed once the new usage rights category has been saved.


## How should a reviewer test this change?

Testing should ensure that;

- Any restrictions and special instructions defined in configuration display in the grid view, image view and upload page as expected for the correct rights category and that all other rights categories are unaffected.
- Any additional restrictions and special instructions defined against an image are added to the diaply after the config derived restrictions and special instructions
- Editing of the rights category for an image results in the corret config derived infomration displaying.
- The behaviour of the uploads page is correct and shows the correct restrictions and special instructions as required and that upload continue to operate as required.
- No other functionality is affected by these changes

## Who should look at this?
<!-- Reach the team with @guardian/digital-cms -->

## Tested? Documented?
- [x ] locally by committer
- [x] locally by Guardian reviewer
- [x] on the Guardian's TEST environment
- [ ] relevant documentation added or amended (if needed)

AndyKilmory commented on kahuna/public/js/components/gr-image-cost-message/gr-image-cost-message.js:
```diff
@@ -1,21 +1,55 @@
+          return "";
+        }
+        let rtxt = "";
+        if (this.image.data.usageRights.usageRestrictions && this.image.data.usageRights.usageRestrictions.length > 0) {
```
Fixed

AndyKilmory commented on kahuna/public/js/preview/image.js:
```diff
@@ -136,6 +136,27 @@ image.controller('uiPreviewImageCtrl', [
+        if (!this.image.data.usageRights) {
+          return "";
+        }
+        let rtxt = "";
```
Not really a utility function space within the current framework and given it only appears in the two locations I think I'm okay with it staying put at present

AndyKilmory commented on kahuna/public/js/upload/jobs/required-metadata-editor.html:
```diff
@@ -109,7 +109,33 @@
-            <div class="job-info--editor__label text-small">Special Instructions</div>
+            <div class="job-info--editor__label text-small job-additional-special-instructions-label">Special Instructions</div>
+
+            <div ng-if="!!ctrl.metadata.usageInstructions">
```
Yes - ng-if will only yake a boolean as a parameter and ctrl.metadata.usageInstructions will be text if it exists - first ! forces truthy evaluation of the parameter which turns it into a boolean - but its the NOT of that value that I want - hence the second !

AndyKilmory commented on kahuna/public/js/usage-rights/usage-rights-editor.html:
```diff
@@ -50,14 +50,27 @@
-            <!-- TODO: potentially pull restrictions out from the
-            server request and add it to all properties -->
+            <!-- TODO: potentially pull restrictions out from the server request and add it to all properties -->
+            <div ng-if="property.name === 'restrictions' && !!ctrl.category.usageRestrictions && ctrl.usageRights.length == 1">
```
as above

AndyKilmory commented on kahuna/public/js/usage-rights/usage-rights-editor.html:
```diff
@@ -50,14 +50,27 @@
                            ng-model="ctrl.showRestrictions"
                            ng-disabled="ctrl.forceRestrictions" />
-                    Restricted
+                    <span ng-if="!!!ctrl.category.usageRestrictions">
+                        Restricted
+                    </span>
+                    <span ng-if="!!ctrl.category.usageRestrictions">
```
Triple ! redcued to single.

Conalb97 commented on kahuna/public/js/components/gr-image-cost-message/gr-image-cost-message.js:
```diff
@@ -1,21 +1,55 @@
+        if (!this.image.data.usageRights) {
+          return "";
+        }
+        let rtxt = "";
```
Looks good to me :)

paperboyo commented:
I think this is very valuable, thank you! Initial functionality/UI feedback:

1. As noted in description, usageInstructions appears on the upload page **only after** Usage Rights is committed. This is confusing. Could we possibly reuse UX from Templates, whereby upon mere selection of a Usage Right with usageInstructions applied, automated Special instructions bit appears without the need to commit the save?
2. Currently usageInstructions use colour `yellow` (we don’t use it anywhere I don’t think) with a hollow border around them and usageRestrictions use a filled background of #ffbc01:
    <img width="847" alt="image" src="https://github.com/guardian/grid/assets/6032869/0f0c93c2-5535-4b3d-8ec2-4c80fcdcbe86">
    This is inconsistent. Could we maybe find a design that makes it clear they are being prepended automatically? (this also applies to how they are presented in the Info panel? It may be something simple, like displaying them in italics, but I defer to designers here.
3. Could we move the label Special Instructions up, so it’s clear those automatically added are part of Special Instructions (see where label sith for eg. Leases):
    <img width="542" alt="image" src="https://github.com/guardian/grid/assets/6032869/3d30c14e-4676-47c3-9662-9ccb058ee2ea">
4. `I want to add further restrictions` and `Add further special instructions` are a tad wordy and also they may make slightly less sense when the checkbox cannot be unchecked. How about both saying something like eg. `Further restrictions:` and `Further special instructions:`? Shorter. Similar to each other, so suggesting both have stuff added automagically (in spirit of pt. 2)
5. This is slightly complicated. I don’t have strong feelings and eager to hear arguments to the contrary, but I think the following makes most sense. For some Usage Rights, Restrictions (hence Restriction notes) are optional (one can uncheck the checkbox) – eg. Obituary. If one adds manual Restriction note to such an image and later unchecks Restricted, this note disappears (and doesn’t come back if you check the box again). Basically, such an image isn’t Restricted any more, so has no Restriction notes. I  think that should also happen for automated Restriction notes – when usageRestrictions is configured for a specific Usage Right, but user unchecks Restricted, automated Restriction note should also disappear as **effectively** the image is no longer Restricted (no orange bit, no icon etc).

Have to run now, will be playing with it more. Exciting!

paperboyo commented:
For future travellers, slightly related to: https://github.com/guardian/grid/pull/3880.

Also, given pt. 5, I wonder if it should not be `Further restrictions:`, but just `Restrictions:` in pt. 4? 🤔

AndyKilmory commented:
Re. point 5 - talking with the BBC team it was felt that if thechosen rights category conferred restrictions on an IMage then the user should not be at liberty to remove those restrictions (certainly in the case of the Programme Images reights categories for the BBC - the restrictions are not optional or for the user to remove) - they can still clarify the restrictions by adding further detail, but the base restrictions are fixed by the rights category. Perhaps if there is a category that has default restrictions that may or may not apply we should use the 'default' property within the rights category restrictions. It was felt that the restrictions conferred in this way were not a property of the image but of the rights category itself - a more traditional normalised data model - and as such shouldn't be stored on the image but held against the rights category.

AndyKilmory commented:
Commits from 11th June and 12th June address points 1 to 4 from above comment

rebecca-thompson commented on kahuna/public/js/upload/jobs/required-metadata-editor.css:
```diff
@@ -5,3 +5,23 @@
+  padding-bottom: 2px;
+}
+
+.job-additional-special-instructions-label {
```
I think this is unused?

rebecca-thompson reviewed (changes requested):
I don't think we need to duplicate `usageInstructions` and `usageRestrictions` in the image response, as they relate to the usage right itself and are not individual to the image.

We can extract them from `ctrl.categories` directly and then use them in the various places they are needed. Ie:

```
    function updateUsageRightsCategory() {
        let category = ctrl.categories.find(cat => cat.value === ctrl.usageRights.data.category);
        ctrl.usageRestrictions = category.usageRestrictions;
        ctrl.usageSpecialInstructions = category.usageSpecialInstructions;
        ctrl.usageRightsCategory = category && category.name;
        ctrl.showUsageRights = ctrl.usageRightsCategory === undefined;
    }
```

This might simplify the client-side code a little bit and is in line with how we are treating other usage rights-specific fields

AndyKilmory commented on kahuna/public/js/components/gr-image-cost-message/gr-image-cost-message.js:
```diff
@@ -1,21 +1,55 @@
+      const states = imageService(ctrl.image).states;
+      ctrl.messageState = (states.hasRestrictions) ? "conditional" : states.costState;
+
+      ctrl.restrictionsText = () => {
```
Will look to extract into shared js utility module

AndyKilmory commented:
Hi Rebecca, 
Thanks for the comments and guidance. 

I opted to incorporate the usage restrictions and usage instructions into the image response to make the information available close to the image restrictions and image special instructions as they are rendered in pairs in the ui - thus the image response remains complete in terms of rendering the image information (even the information not stored directly on the image) - more of an MVC type pattern with the image response the view providing information for displaying the image rather than needing to reference a secondary collection/model in the UI. This would be helpful if we start integrating the system with other third party systems (which is planned for BBC Images) - a call to the api will provide all the information required to display the image in the third party system - including restrictions guidance - without this information in the image response we'd need additional calls to get this information from the rights categories. 

I appreciate this is a break from patterns used elsewhere, but I think the idea of having information not stored directly on the image being required to fully explain the image is not the usual pattern followed - its more like a traditional relational data model - pulling the required information from the view from multiuple 'tables'/objects.

However, I do understand that in the current model architecture the more natrual home for these data is the rights category model so I will look to implement the requested changes.

AndyKilmory commented:
I have now removed the unused css class; moved the restrictionsText logic into a utility library referenced in both locations. I have, however, retained the additional rights restrictions and instructions in the image response. Looking at the image response in the EditsApi.scala elements of the rights category data are currently incorporated into the Image Response and adding the additional instructions and restrictions does seem consistent - it supports future integrations with 3rd party systems where the full image information will be available in a single call and also supports editing/updating the rights category in the infoPanel on the grid view and the metadata panel in the image view - its only the upload page the image-editor construct and the need to support updates in the other areas means the use of the image-response approach appears more consistent.

rebecca-thompson commented:
> I have now removed the unused css class; moved the restrictionsText logic into a utility library referenced in both locations. I have, however, retained the additional rights restrictions and instructions in the image response. Looking at the image response in the EditsApi.scala elements of the rights category data are currently incorporated into the Image Response and adding the additional instructions and restrictions does seem consistent - it supports future integrations with 3rd party systems where the full image information will be available in a single call and also supports editing/updating the rights category in the infoPanel on the grid view and the metadata panel in the image view - its only the upload page the image-editor construct and the need to support updates in the other areas means the use of the image-response approach appears more consistent.

Hi Andy, just to clarify... are you referring to the edit API endpoint `GET /metadata/:id`? I think this follows the same pattern as the media API image endpoint where it returns the usage category and any restrictions that have been set that are _unique to the image_, ie, they have been entered via the free text field in the metadata panel. In both APIs, `category` acts as a foreign key by which usage-rights-specific metadata (such as description and now instructions and restrictions following this pr) are retrieved. Let me know if I've misunderstood.

Happy to chat about it, if you have some time. I'm not as familiar with this codebase, so it would be great to understand this better.

AndyKilmory commented:
Hi Rebecca,
Apologies - your are correct - in the 'image response' the rights category identifier is included as a foreign key and the restrictions and other information included in that section do directly relate to the image itself - I shouldn't try writing messages late on Friday afternoons! However, I'm still concerned that the removal of the rights specific 'usage restrictions' and 'usage instructions' from the 'image response' will cause problems down the line, and including them there is the better pattern.

Almost all the data fields loaded into the CategoryResponse (EditsApi.scala) are only utilised during the editing/updating of the rights category for an image and don't form part of the 'standard' display of an image - that applies on the uploads page, in the info-panel in the grid view, and in the metadata panel when viewing a single image. The only exception to this is the 'name' field. In the info-panel this is covered by the gr-image-metadata control - which displays the rights category name via the selectedUsageCategory() function;

![Screenshot 2024-06-24 at 16 39 30](https://github.com/guardian/grid/assets/128470622/72ac4130-15a3-4478-950b-bed63577c70d)

This function makes an API call to the editsApi to get all the rights category information which it then filters down to the correct display text (subject to the complexity of multiple images spanning multiple categories having been selected). The same control structure is used in the metadata panel. 

The new usageRestrictions and usageInstructions fields are more akin to the rights category name in the info-panel etc. - they are required to fully display image information and not just used during updating of the rights category (the supplemental elements can be mapped to other metadata fields - but this is already covered by separate calls). The problem I foresee is that if we remove the useageRestrictions and usageInstructions from the imageResponse we will need to make additional calls to the edits api to get these data when displaying the image information. There will also need to be extensions to the gr-image-cost-message control - this displays the image restrictions info. The grid display of each image will also need to be modified as the hover text for the restrictions flag includes the usage restrictions (not sure if that will require an api call per grid image or if we can re-factor so that it's just one additional call). The upload page is probably okay as it assumes the need to update the rights category for images and so has the rights category collection available.

I think this problem will become more widespread as the BBC start to integrate BBC Images with other BBC systems (work planned for this year and into the future) - any request to pull image information from BBC Images will include a call to the media-api to get the image data - which will then require a further call to the EditsApi to get the rights category data so that the 'usage restrictions' and 'usage instructions' can be displayed with the rest of the image data. If this information were available in the 'image response' no further calls will be required. (thinking about it now I wonder if we should also extend the 'images response' to include the name of the rights category to remove the need for the existing edits api call mentioned above?). 

I know that this is a 'new' pattern to the way that data are 'packaged' within BBC Images/The Grid - but I think the data design pattern (abstracting data used in image display out of the image object itself) is not really used elsewhere either. I don't think it impacts search as the information isn't image specific and equivalent results can be obtained by searching on 'category' or any other aspect of the system.

The pattern is effectively a traditional MVC style pattern - we have a distributed data model (image specific data and rights categories) and the image response represents a view of these data - blending the image and rights category data in such a way to provide sufficient information in the view to fully display the image in the UI.

I've included the 'usage instructions' in the main metadata section so it is within the same grouping as the special instructions with which it is paired, but I think now that this should perhaps be in the usageRights section to make it clear it originates from the rights category selection.

Sorry for the length of the response - trying to be clearer and more precise this time! Happy to arrange a meeting so we can chat through these issues in more detail - do you want to suggest a convenient time?

Thanks,
Andy

paperboyo commented:
Thanks for the review, @rebecca-thompson! This awaits some smol UI tweaks discussed off-PR with @AndyKilmory.

paperboyo commented:
@AndyKilmory Thank you for the adjustments, this looks great and is ready to merge (probably tomorrow). The minorest of minor things is that in Info panel, the line height for `job-usage-instructions` is 24px, while it should be 19.5px as the rest of text:

![image](https://github.com/guardian/grid/assets/6032869/d951d8e7-002f-47ad-876d-c4f2186108da)

vs. everywhere else:

![image](https://github.com/guardian/grid/assets/6032869/e1ab605e-fc12-49e9-9efc-b3ab72d6b999)

paperboyo commented:
Ouch, apologies, I spoke too soon! It appears I cannot add or edit Special Instructions (in both Browser and Viewer) whether image has them or not, having correct edit_metadata permission. Could you take a look, please?

Adding:
![spec](https://github.com/guardian/grid/assets/6032869/48df1c27-ffe8-4088-96d8-90c7596ea67c)

Editing:
![spec2](https://github.com/guardian/grid/assets/6032869/39ff7693-0371-4d6c-8088-0584f5411f1d)

AndyKilmory commented:
Thanks for the spot - will take a look

AndyKilmory commented:
Fix for Special Instructions save error completed

AndyKilmory commented:
Have added additional code to ensure the Description and Special INstructions textareas are responsive to keyboard input for cancel (Escape) and save (Ctrl/Cmd + Enter) and also that cursor starts at end of content when starting edit.

paperboyo commented:
Thank you for going the extra mile, @AndyKilmory! Great to see Description finally behaving like other fields with ESC/Cmd+Enter! We will give it another test and, hopefully, merge soon.
